### PR TITLE
allowing tests to run without the play context

### PR DESCRIPTION
**Pull request by eishay**
Loading the play context (plugins, configs, globals etc) is very expensive. This method allows us to cut 30%-50% of the test time if the test involves db and even more if it just needs guice injection.
I will check if we can trim some more time by reusing tables instead of recreating them.

Some examples:
**before**
BookmarkTest
[info] + load all (2 seconds, 941 ms)
[info] + load by user (579 ms)
[info] + load by uri (414 ms)
[info] + count all (361 ms)
[info] + count by user (358 ms)
[info] + count mutual keeps (334 ms)
[info] Passed: : Total 6, Failed 0, Errors 0, Passed 6, Skipped 0
[success] Total time: 7 s, completed Apr 25, 2013 12:32:00 AM
**after**
info] + load all (1 second, 931 ms)
[info] + load by user (167 ms)
[info] + load by uri (139 ms)
[info] + count all (91 ms)
[info] + count by user (118 ms)
[info] + count mutual keeps (140 ms)
[info] Passed: : Total 6, Failed 0, Errors 0, Passed 6, Skipped 0
[success] Total time: 5 s, completed Apr 25, 2013 12:38:28 AM

NormalizedURITest
**before**
info] bookmark pagination should
[info] + get all (2 seconds, 895 ms)
[info] + get first (554 ms)
[info] + get last (397 ms)
[info] + get none (381 ms)
[info] get by state should
[info] + search gets nothing (408 ms)
[info] NormalizedURIs search by url should
[info] + search gets nothing (285 ms)
[info] + search gets short (283 ms)
[info] + search gets long (333 ms)
[info] NormalizedURIs get created url should
[info] + search gets nothing (273 ms)
[info] + search gets short (295 ms)
[info] Passed: : Total 10, Failed 0, Errors 0, Passed 10, Skipped 0
[success] Total time: 8 s, completed Apr 25, 2013 12:42:01 AM
**after**
[info] bookmark pagination should
[info] + get all (1 second, 978 ms)
[info] + get first (148 ms)
[info] + get last (149 ms)
[info] + get none (114 ms)
[info] get by state should
[info] + search gets nothing (191 ms)
[info] NormalizedURIs search by url should
[info] + search gets nothing (81 ms)
[info] + search gets short (91 ms)
[info] + search gets long (78 ms)
[info] NormalizedURIs get created url should
[info] + search gets nothing (64 ms)
[info] + search gets short (64 ms)
[info] Passed: : Total 10, Failed 0, Errors 0, Passed 10, Skipped 0
[success] Total time: 5 s, completed Apr 25, 2013 12:52:48 AM

### CODE REVIEW

File: shoebox/app/com/keepit/common/db/slick/H2.scala [hunk 1]
```diff
@@ -3,12 +3,23 @@ package com.keepit.common.db.slick
...
+trait TableInitListener {
+  def init(table: TableWithDDL): Unit
+}
 
 // see https://groups.google.com/forum/?fromgroups=#!topic/scalaquery/36uU8koz8Gw
 class H2(val dbInfo: DbInfo)
     extends DataBaseComponent {
   println("initiating H2 driver")
   val Driver = H2Driver
+  val tablesToInit = new HashMap[String, TableWithDDL]
```
Comment (2is10) [inline shoebox/app/com/keepit/common/db/slick/H2.scala]: should this be threadsafe?

File: shoebox/app/com/keepit/common/db/slick/H2.scala [hunk 2]
```diff
@@ -21,6 +32,19 @@ class H2(val dbInfo: DbInfo)
   }
 
   override def entityName(name: String): String = name.toUpperCase()
+
+  override def tableToInit(table: TableWithDDL) {
```
Comment (2is10) [inline shoebox/app/com/keepit/common/db/slick/H2.scala]: verb names for methods really help me understand them easier. how about `initTableNowOrLater`? or `initTable` and rename the private one to `initTableNow`?

File: shoebox/test/com/keepit/test/TestDBRunner.scala [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+  def withDB[T](overrideingModules: Module*)(f: RichInjector => T) = {
+    withInjector(overrideingModules: _*) { implicit injector =>
+      val db = inject[DataBaseComponent]
+      val h2 = inject[DataBaseComponent].asInstanceOf[H2]
+      h2.initListener = Some(new TableInitListener {
+        def init(table: TableWithDDL) = initTable(db, table)
+      })
+      try {
+        (f(injector))
```
Comment (2is10) [inline shoebox/test/com/keepit/test/TestDBRunner.scala:50]: why the outer parens?

File: shoebox/test/com/keepit/test/TestDBRunner.scala [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+  def socialUserInfoRepo(implicit injector: RichInjector) = inject[SocialUserInfoRepo]
+  def installationRepo(implicit injector: RichInjector) = inject[KifiInstallationRepo]
+  def userExperimentRepo(implicit injector: RichInjector) = inject[UserExperimentRepo]
+  def emailAddressRepo(implicit injector: RichInjector) = inject[EmailAddressRepo]
+  def unscrapableRepo(implicit injector: RichInjector) = inject[UnscrapableRepo]
+
+  def withDB[T](overrideingModules: Module*)(f: RichInjector => T) = {
+    withInjector(overrideingModules: _*) { implicit injector =>
+      val db = inject[DataBaseComponent]
+      val h2 = inject[DataBaseComponent].asInstanceOf[H2]
```
Comment (2is10) [inline shoebox/test/com/keepit/test/TestDBRunner.scala:45]: `db.asInstanceOf[H2]`

File: shoebox/test/com/keepit/test/TestInjector.scala [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+trait TestInjector {
+
+  def inject[A](implicit m: Manifest[A], injector: RichInjector): A = injector.inject[A]
+
+  def withInjector[T](overrideingModules: Module*)(f: RichInjector => T) = {
+    def dbInfo: DbInfo = TestDbInfo.dbInfo
+    DriverManager.registerDriver(new play.utils.ProxyDriver(Class.forName("org.h2.Driver").newInstance.asInstanceOf[Driver]))
+//    val conn = DriverManager.getConnection(TestDbInfo.url)
+    val modules = {
+      def overridModule(m: Module, overriding: Module) = Modules.`override`(Seq(m): _*).`with`(overriding)
```
Comment (2is10) [inline shoebox/test/com/keepit/test/TestInjector.scala:29]: `overrideModule` ? (spelling)

File: shoebox/test/com/keepit/test/TestInjector.scala [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+import java.sql.DriverManager
+import java.sql._
+import javax.sql._
+import akka.actor.ActorSystem
+
+trait TestInjector {
+
+  def inject[A](implicit m: Manifest[A], injector: RichInjector): A = injector.inject[A]
+
+  def withInjector[T](overrideingModules: Module*)(f: RichInjector => T) = {
```
Comment (2is10) [inline shoebox/test/com/keepit/test/TestInjector.scala:24]: `overridingModules` (spelling, and inTestDBRunner.scala)

**Comment by 2is10** [comment]
wow, impressive, great idea!

File: shoebox/app/com/keepit/common/db/slick/H2.scala (hunk 1 above)
Comment (eishay) [inline shoebox/app/com/keepit/common/db/slick/H2.scala]: No, though I don't think we're opening threads in tests. Afaik, we're single threaded (per test).

Comment (eishay) [inline shoebox/app/com/keepit/common/db/slick/H2.scala]: actually, the table may be populated in dev mode. i'll make it threadsafe

File: shoebox/app/com/keepit/common/db/slick/H2.scala (hunk 2 above)
Comment (eishay) [inline shoebox/app/com/keepit/common/db/slick/H2.scala]: sure

File: shoebox/test/com/keepit/test/TestDBRunner.scala (hunk 3 above)
Comment (eishay) [inline shoebox/test/com/keepit/test/TestDBRunner.scala:50]: no need

File: shoebox/test/com/keepit/test/TestDBRunner.scala (hunk 4 above)
Comment (eishay) [inline shoebox/test/com/keepit/test/TestDBRunner.scala:45]: sure, actually don't need db from the first place

**Pull request merged**